pkg/controller/gcppubsub/reconcile.go

```diff
@@ -0,0 +1,328 @@
+}
+
+func (r *reconciler) getLabelSelector(src *v1alpha1.GcpPubSubSource) labels.Selector {
+	ls := labels.NewSelector()
+	for k, v := range getLabels(src) {
+		req, err := labels.NewRequirement(k, selection.Equals, []string{v})
+		if err != nil {
+			// Probably fail.
+		}
+		ls.Add(*req)
+	}
+	return ls
+}
+
+func getLabels(src *v1alpha1.GcpPubSubSource) map[string]string {
+	return map[string]string{
+		"knative-eventing-source":      controllerAgentName,
+		"knative-eventing-source-name": src.Name,
+	}
+}
+
+func (r *reconciler) makeReceiveAdapter(src *v1alpha1.GcpPubSubSource, subscriptionId, sinkURI string) *v1.Deployment {
+	credsVolume := "google-cloud-key"
+	credsMountPath := "/var/secrets/google"
+	credsFile := fmt.Sprintf("%s/key.json", credsMountPath)
+	replicas := int32(1)
+	dLabels := getLabels(src)
+	return &v1.Deployment{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace:    src.Namespace,
+			GenerateName: fmt.Sprintf("gcppubsub-%s-", src.Name),
+			Labels:       dLabels,
+			OwnerReferences: []metav1.OwnerReference{
+				{
+					APIVersion: v1alpha1.SchemeGroupVersion.String(),
+					Kind:       "GcpPubSubSource",
+					Name:       src.Name,
+					UID:        src.UID,
+					Controller: &trueVal,
+				},
+			},
+		},
+		Spec: v1.DeploymentSpec{
+			Selector: &metav1.LabelSelector{
+				MatchLabels: dLabels,
+			},
+			Replicas: &replicas,
+			Template: corev1.PodTemplateSpec{
+				ObjectMeta: metav1.ObjectMeta{
+					Annotations: map[string]string{
+						"sidecar.istio.io/inject": "true",
+					},
+					Labels: dLabels,
+				},
+				Spec: corev1.PodSpec{
+					ServiceAccountName: r.receiveAdapterServiceAccountName,
+					Containers: []corev1.Container{
+						{
+							Name:  "receive-adapter",
+							Image: r.receiveAdapterImage,
+							Env: []corev1.EnvVar{
+								{
+									Name:  "GOOGLE_APPLICATION_CREDENTIALS",
+									Value: credsFile,
+								},
+								{
+									Name:  "GCPPUBSUB_PROJECT",
+									Value: src.Spec.GoogleCloudProject,
+								},
+								{
+									Name:  "GCPPUBSUB_TOPIC",
+									Value: src.Spec.Topic,
+								},
+								{
+									Name:  "GCPPUBSUB_SUBSCRIPTION_ID",
+									Value: subscriptionId,
+								},
+								{
+									Name:  "SINK_URI",
+									Value: sinkURI,
+								},
+							},
+							VolumeMounts: []corev1.VolumeMount{
+								{
+									Name:      credsVolume,
+									MountPath: credsMountPath,
+								},
+							},
+						},
+					},
+					Volumes: []corev1.Volume{
+						{
+							Name: credsVolume,
+							VolumeSource: corev1.VolumeSource{
+								Secret: &corev1.SecretVolumeSource{
+									SecretName: src.Spec.GcpCredsSecret,
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+func (r *reconciler) createSubscription(ctx context.Context, src *v1alpha1.GcpPubSubSource) (*pubsub.Subscription, error) {
+	// Auth to GCP is handled by having the GOOGLE_APPLICATION_CREDENTIALS environment variable
+	// pointing at a credential file.
```

Comment:
Are the credentials for creating a subscription shared by all sources, or are they source specific?

Reply:
There are two auth credentials:

1. The credentials used to create/delete the Subscription. 
    - Used only by the GCP PubSub Source controller (this file).
    - There is a single credential for all sources.
2. The credentials used to pull the Subscription.
    - Used only by the Receive Adapters.
    - Each source may use a distinct credential. It is specified in the spec (spec.gcpCredsSecret).

I can see the argument for having 1 also be per Source, but didn't want to tackle that complexity now. So the Cluster admin is giving fairly wide permissions to create/delete subscriptions by just installing the GcpPubSubSource controller, _but_ the only way to read the messages is to have access to a usable secret in your namespace.